frontend/packages/substrate-react/src/dotsama/DotSamaContext.tsx

```diff
@@ -77,47 +102,61 @@ export const DotSamaContextProvider = ({
     [chainId in RelayChainId]: RelaychainApi;
   }>(RELAYCHAIN_PROVIDERS_DEFAULT);
 
-  const [extensionStatus, setExtensionStatus] = useState<DotSamaExtensionStatus>('initializing');
-
+  const [extensionInjected, setInjectedExtension] = useState<
+    InjectedExtension | undefined
+  >(undefined);
+  const [extensionStatus, setExtensionStatus] =
+    useState<DotSamaExtensionStatus>("initializing");
   const activate = async (
-    selectDefaultAccount: boolean = true
-  ): Promise<any[] | undefined> => {
+    walletId: SupportedWalletId = "polkadot-js",
+    selectDefaultAccount: boolean = false
+  ): Promise<any | undefined> => {
     setExtensionStatus("connecting");
 
-    let extensionExists = true;
-    let inectedExtesions;
+    let injectedExtension, extensionError;
     try {
-      const extensionPkg = await import("@polkadot/extension-dapp");
-      inectedExtesions = await extensionPkg.web3Enable(appName);
-      extensionExists = inectedExtesions.length !== 0;
+      if (!window.injectedWeb3) throw new Error("Extension not installed.");
+
+      let extension = window.injectedWeb3[walletId];
+      if (!extension) throw new Error("Extension not installed.");
+
+      injectedExtension = (await extension.enable(
+        appName
+      )) as InjectedExtension;
     } catch (e) {
       console.error(e);
-      extensionExists = false;
+      extensionError = e;
     }
 
-    if (!extensionExists) {
-      setExtensionStatus('no_extension');
-      return inectedExtesions;
+    if (injectedExtension === undefined) {
+      setExtensionStatus("no_extension");
+      return Promise.reject(extensionError);
     }
 
-    setExtensionStatus('connected');
+    setExtensionStatus("connected");
+    localStorage.setItem("wallet-id", walletId);
+    setInjectedExtension(injectedExtension);
 
     for (let i = 0; i < supportedParachains.length; i++) {
       const { chainId } = supportedParachains[i];
       const { prefix } = ParachainNetworks[chainId];
 
       try {
-        const extensionPkg = await import("@polkadot/extension-dapp");
-        const accounts = await extensionPkg.web3Accounts({
-          ss58Format: prefix,
-        });
+        let accounts = await injectedExtension.accounts.get();
+        if (accounts === undefined)
+          throw new Error("Unable to fetch accounts from extension.");
+        accounts = mapAccounts(walletId, accounts, prefix);
+        if (accounts === undefined)
+          throw new Error("Unable to fetch accounts from extension.");
 
         setParachainProviders((s) => {
-          s[chainId].accounts = accounts.map((x, i) => ({
-            address: x.address,
-            name: x.meta.name ?? i.toFixed(),
-          }));
-          return { ... s };
+          s[chainId].accounts = (accounts as InjectedAccountWithMeta[]).map(
+            (x, i) => ({
+              address: x.address,
+              name: x.meta.name ?? i.toFixed(),
```

Comment:
Question: instead of index as fallback, do you think it would be better for end user if we use the shortened address instead? something like:

`5Grwva..KutQY`

Reply:
@easteregg can be done, it will look much better than a number.